### PR TITLE
Filetypedfilenameinput

### PR DESCRIPTION
**Pull request by fatoldsun00**
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the [contribution guidelines](https://github.com/node-red/node-red/blob/master/CONTRIBUTING.md)
- [x] For non-bugfix PRs, I have discussed this change on the forum/slack team.
- [x] I have run `grunt` to verify the unit tests pass
- [ ] I have added suitable unit tests to cover the new/changed functionality

### CODE REVIEW

**Comment by dceejay** [comment]
Hi Is this related to this discussion ? https://discourse.nodered.org/t/pr-add-typed-input-for-filename-prop-on-native-file-in-node/24253/14

**Comment by dceejay** [comment]
You also need to handle - around line 151 in the .js file - the case where the node.filename for fileout  is now not a static entity - as we have to close it after every write in case it changes between writes. Currently this is just checking if filename is defined... - now it will have to check the type and if not str or env it will have to close. (Which does affect performance so should also be noted in the info sidebar.

File: packages/node_modules/@node-red/nodes/core/storage/10-file.js [hunk 1]
```diff
@@ -257,7 +259,8 @@ module.exports = function(RED) {
         var node = this;
 
         this.on("input",function(msg, nodeSend, nodeDone) {
-            var filename = (node.filename || msg.filename || "").replace(/\t|\r|\n/g,'');
+            var filename = node.filename ? RED.util.evaluateNodeProperty(node.filename, node.filenameType, msg, msg) : (msg.filename || "");
+            var filename = filename.replace(/\t|\r|\n/g,'');
```
[the remark this answers is not in the document]
Comment (fatoldsun00) [reply packages/node_modules/@node-red/nodes/core/storage/10-file.js]: sry

File: packages/node_modules/@node-red/nodes/core/storage/10-file.js [hunk 2]
```diff
@@ -50,7 +51,7 @@ module.exports = function(RED) {
         node.closeCallback = null;
 
         function processMsg(msg,nodeSend, done) {
-            var filename = node.filename || msg.filename || "";
+            var filename = node.filename ? RED.util.evaluateNodeProperty(node.filename, node.filenameType, msg, msg) : (msg.filename || "");
```
[the remark this answers is not in the document]
Comment (fatoldsun00) [reply packages/node_modules/@node-red/nodes/core/storage/10-file.js:54]: Thank for reply, indeed evaluateNodeProperty take a callback.

Can i wrap evaluateNodePorperty like this 

`function evaluateNodeProperty(input,inputType,node,msg){
             return new Promise(function (resolve, rejected) {
                try {
                    RED.util.evaluateNodeProperty(input, inputType, node, msg,(err,rep)=>{
                        if (err) throw (err)
                        resolve(rep)
                    })
                } catch (err) {
                    rejected(err)
                } 
             })
        }`

and change processMsg for Async fn like this 

`async function processMsg(msg,nodeSend, done){`

Then i can await evaluateNodeProperty

It's alright or there thing that i miss ?

**Comment by fatoldsun00** [comment]
@dceejay Alright, if we append to file and type is not str or env, the filename can change.

Then i can overwrite if (node.filename) with if (node.filename && ['str','env'].indexOf(node.filenameType) > -1)

But in this case we close stream even filename doesn't change... I guess i need to save the previsous filename and compare it to the filename IF type !str && !env in perf terms

**Comment by dceejay** [comment]
well we can't do that comparison after we have written - and we don't really want to close and reopen right at the start of a write (I think) - plus it's quite handy to close files after we do write for some OS like windows so they can then read them... so I think it is safer just to close if there is a possibility it may change for now. 

I think it would need some tests to check how good or bad it would be to a) not close at all after write (especially on Windows) - and then b) to do the check before write and close and reopen and then write especially as these are all async.

**Comment by fatoldsun00** [comment]
I try to make time soon for some perf test

**Pull request closed without merging**